Swift 5.8 release blog post (#257)
apple/swift-org-website: _posts/2023-03-17-swift-5.8-released.md

To take things even further, you can now [customise the appearance of your documentation pages](https://www.swift.org/documentation/docc/customizing-the-appearance-of-your-documentation-pages) with colour, font, and icon customisations. Navigation also took a step forward with quick navigation, allowing fuzzy in-project search:

![A DocC documentation page showing a quick navigation overlay showing fuzzy documentation search](/assets/images/5.8-blog/docc-fuzzy-search.png)

> Swift-DocC also now supports documenting extensions to types from other modules. This is an opt-in feature, and can be [enabled by adding the `--include-extended-types` flag when using the Swift-DocC plugin](https://apple.github.io/swift-docc-plugin/documentation/swiftdoccplugin/generating-documentation-for-extended-types).

Author: Note: This link currently leads to a 404 as the content is pending deployment.

- SE-0369: [Add CustomDebugStringConvertible conformance to AnyKeyPath](https://github.com/apple/swift-evolution/blob/main/proposals/0369-add-customdebugdescription-conformance-to-anykeypath.md)
- SE-0370: [Pointer Family Initialization Improvements and Better Buffer Slices](https://github.com/apple/swift-evolution/blob/main/proposals/0370-pointer-family-initialization-improvements.md)
- SE-0372: [Document Sorting as Stable](https://github.com/apple/swift-evolution/blob/main/proposals/0372-document-sorting-as-stable.md)
- SE-0373: [Lift all limitations on variables in result builders](https://github.com/apple/swift-evolution/blob/main/proposals/0373-vars-without-limits-in-result-builders.md)
- SE-0374: [Add sleep(for:) to Clock](https://github.com/apple/swift-evolution/blob/main/proposals/0374-clock-sleep-for.md)
- SE-0375: [Opening existential arguments to optional parameters](https://github.com/apple/swift-evolution/blob/main/proposals/0375-opening-existential-optional.md)
- SE-0376: [Function Back Deployment](https://github.com/apple/swift-evolution/blob/main/proposals/0376-function-back-deployment.md)

> - (?) SE-0380: [if and switch expressions](https://github.com/apple/swift-evolution/blob/main/proposals/0380-if-switch-expressions.md)

Author: SE-0380 is marked as implemented in Swift 5.9, but it seems to have been cherry-picked into 5.8 in [this PR](https://github.com/apple/swift/pull/63378). Maybe @hborla can clarify?

Reviewer: SE-0380 was reverted from the `release/5.8` branch in https://github.com/apple/swift/pull/63649. It's only implemented in Swift 5.9

> - (?) SE-0381: [DiscardingTaskGroups](https://github.com/apple/swift-evolution/blob/main/proposals/0381-task-group-discard-results.md)

Author: SE-0381 seems to have been [cherry-picked into Swift 5.8](https://github.com/apple/swift/pull/62963), but it's still marked as "Accepted". @hborla do we need to update the evolution status?

Reviewer 2: Discarding groups are accepted but didn’t make it into the SDKs… they ARE available on Linux toolchains though. We should explain this here a bit.

Author: I simply removed SE-0381 from the list for now, but feel free to suggest a way to list it without confusing people about the availability. 🙂

To take things even further, you can now [customise the appearance of your documentation pages](https://www.swift.org/documentation/docc/customizing-the-appearance-of-your-documentation-pages) with colour, font, and icon customisations. Navigation also took a step forward with quick navigation, allowing fuzzy in-project search:

![A DocC documentation page showing a quick navigation overlay showing fuzzy documentation search](/assets/images/5.8-blog/docc-fuzzy-search.png)

Swift-DocC also now supports documenting extensions to types from other modules. This is an opt-in feature, and can be [enabled by adding the `--include-extended-types` flag when using the Swift-DocC plugin](https://apple.github.io/swift-docc-plugin/documentation/swiftdoccplugin/generating-documentation-for-extended-types).

![A documentation page featuring an extension to the standard library's Collection type.](/assets/images/5.8-blog/docc-extended-type.png)

### Swift Package Manager

> Here are some of the changes that have been introduced to [Swift Package Manager](https://github.com/apple/swift-package-manager) in Swift 5.8:

Reviewer 3: ```suggestion
Following are some highlight from the changes introduced to the [Swift Package Manager](https://github.com/apple/swift-package-manager) in Swift 5.8:
```

Author: 👍 I think `highlight` -> `highlights` though?

Reviewer 3: updated

---
layout: post
published: true
date: 2023-03-17 10:00:00 # TODO: Update with official publishing date
title: Swift 5.8 Released!
author: [alexandersandberg]
---

> Swift 5.8 is now officially released! 🎉 This release includes major additions to the [language and standard library](#language-and-standard-library) including piecemeal adoption of upcoming features, an improved [developer experience](#developer-experience), improvements to tools in the Swift ecosystem including [Swift-DocC](#swift-docc), [Swift Package Manager](#swift-package-manager), and [SwiftSyntax](#swiftsyntax), refined [Windows support](#windows-platform), and more.

Author: I'm not sure if we want to add these anchor links to the below sections here or not. What do we think?

Reviewer 4: Maybe useful in the future, worth keeping IMO.

> Swift 5.8 is now officially released! 🎉 This release includes major additions to the [language and standard library](#language-and-standard-library) including piecemeal adoption of upcoming features, an improved [developer experience](#developer-experience), improvements to tools in the Swift ecosystem including [Swift-DocC](#swift-docc), [Swift Package Manager](#swift-package-manager), and [SwiftSyntax](#swiftsyntax), refined [Windows support](#windows-platform), and more.

Reviewer 5: ```suggestion
Swift 5.8 is now officially released! 🎉 This release includes major additions to the [language and standard library](#language-and-standard-library), including `hasFeature` to support piecemeal adoption of upcoming features, an improved [developer experience](#developer-experience), improvements to tools in the Swift ecosystem including [Swift-DocC](#swift-docc), [Swift Package Manager](#swift-package-manager), and [SwiftSyntax](#swiftsyntax), refined [Windows support](#windows-platform), and more.
```

A naive reading of the text as-is suggests that it's the Swift 5.8 release that adopts upcoming features piecemeal (a puzzling state of affairs, although I guess with the `clock(for:)` issue that's also technically true), rather than that it ships with a feature to allow _users_ to adopt upcoming features piecemeal. This wouldn't be confusing for those who've followed Swift Evolution closely, but I'd like to think our audience for this post would be broader than that.

Thank you to everyone in the Swift community who made this release possible. Your Swift Forums discussions, bug reports, pull requests, educational content, and other contributions are always appreciated!

For a quick dive into some of what’s new in Swift 5.8, check out this [playground](https://github.com/twostraws/whats-new-in-swift-5-8) put together by Paul Hudson.

> [The Swift Programming Language](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/) book has been updated for Swift 5.8 and is now published with DocC. This is the official Swift guide and a great entry point for those new to Swift. The Swift community also maintains a number of [translations](/documentation/#translations).

Reviewer 6: Is it also worth noting that TSPL is open for contributions?  We already highlighted that in a [blog post](https://www.swift.org/blog/tspl-on-docc/) for the first beta version that was built using DocC.

Author: I think this would be great, but maybe in the Swift-DocC section instead where this is already talked about in more detail?

Reviewer 6: To me, it doesn't quite seem like the right organization for the DocC section to discuss contributing to TSPL.  TSPL is a project that uses DocC, but it's not part of DocC.

We could link to the README or the CONTRIBUTING file in the book's repo.

https://github.com/apple/swift-book/blob/main/CONTRIBUTING.md

Author: I do agree, but as the open-sourcing of TSPL is [talked about in the DocC section](https://github.com/apple/swift-org-website/pull/257/files#diff-53d76ecf3c343ac9892cbd55408c09237cca4bc4e86b7f16f4f92df9b763d876R81), I think it might still be a suitable place.

## Language and Standard Library

> Swift 5.8 enables you to start incrementally preparing your projects for Swift 6 using _upcoming features_. Upcoming features are used with a new compiler flag `-enable-upcoming-feature FeatureIdentifier` where `FeatureIdentifier` is the name of the specific feature to enable.

Reviewer 6: ```suggestion
Swift 5.8 enables you to start incrementally preparing your projects for Swift 6 using _upcoming features_. By default, upcoming features are disabled. To enable a feature, pass the compiler flag `-enable-upcoming-feature` followed by feature's identifier.
```

Avoiding passive voice and making it clearer that these are off unless you ask for them.

Author: Great suggestion! I feel like a "the" is missing though: "...followed by _the_ feature's..."

Reviewer 7: Is it important that readers know that an upcoming feature is a source-breaking language change coming in Swift 6 that they can adopt immediately?

Author: Maybe it's enough to link out to [the proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0362-piecemeal-future-features.md) for those who want to learn more?

Reviewer 7: Yes, you are probably right. It is probably impossible to capture all its subtleties in a short blurb.

Author: I added a link in https://github.com/apple/swift-org-website/pull/257/commits/bf4e3ed3c2b97d060de1ab0d0dcacc72e0952f3c.

Feature identifiers can also be used in source code using `#if hasFeature(FeatureIdentifier)` so that code can still compile with older tools where the upcoming feature is not available:

```swift
#if compiler(>=5.7) && hasFeature(BareSlashRegexLiterals)
let regex = /.../
#else

> let regex = try NSRegularExpression(pattern: "...")

Author: From my understanding of the [examples in the proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0362-piecemeal-future-features.md#feature-detection-in-source-code), `#/.../#` requires Swift 5.7 or above, which means it's not available in the `#else` here.

> let regex = try NSRegularExpression(pattern: "...")

Reviewer 5: Good point; it should still be available if using Swift 5.8 (as is the implication here) but users who want code to compile in earlier versions wouldn't have that option. This brings me to another point: to use `hasFeature`, one would have to gate on `compiler(>=5.8)`, no? So perhaps this example ought to be:

```swift
#if compiler(>=5.8) && hasFeature(BareSlashRegexLiterals)
let regex = /.../
#elseif compiler(>= 5.7)
let regex = #/.../#
#else
let regex = try NSRegularExpression(pattern: "...")
#endif
```

> let regex = try NSRegularExpression(pattern: "...")

Reviewer 8: I don't think we should have the `#else`. For a start, the code won't really work as it implies because `NSRegularExpression` is not a swift regular expression.

I suggest just 

```
#if compiler(>=5.8) && hasFeature(BareSlashRegexLiterals)
let regex = /.../
#else
let regex = #/.../#
#endif
```

with code that assumes minimum tooling of 5.7.

> let regex = try NSRegularExpression(pattern: "...")

Reviewer 8: (this is the norm fwiw, most code is written to work with the current and current-1 tools)

> let regex = try NSRegularExpression(pattern: "...")

Reviewer 8: I removed the example – it's actually not even valid because the short-circuiting `&&` was only introduced in 5.8 so earlier tools can't handle it.

> let regex = try NSRegularExpression(pattern: "...")

Author: 👍 Probably best, to not raise any confusion. There are examples in the proposal for those interested.

#endif
```

Swift 5.8 includes upcoming features for the following Swift evolution proposals:

- SE-0274: [Concise magic file names](https://github.com/apple/swift-evolution/blob/main/proposals/0274-magic-file.md) (`ConciseMagicFile`)
- SE-0286: [Forward-scan matching for trailing closures](https://github.com/apple/swift-evolution/blob/main/proposals/0286-forward-scan-trailing-closures.md) (`ForwardTrailingClosures`)
- SE-0335: [Introduce existential any](https://github.com/apple/swift-evolution/blob/main/proposals/0335-existential-any.md) (`ExistentialAny`)

> - SE-0354: [Regex literals](https://github.com/apple/swift-evolution/blob/main/proposals/0354-regex-literals.md) (`BaseSlashRegexLiterals`)

Reviewer 5: We expect that Objective-C forward declarations will have an "upcoming feature" designation shortly in time for Swift 5.8.

For example, building the following file at `/Users/example/Desktop/0274-magic-file.swift` in a module called `MagicFile` with `-enable-experimental-feature ConciseMagicFile` will opt into the concise format for `#file` and `#filePath` described in SE-0274:

```swift
print(#file)
print(#filePath)
fatalError("Something bad happened!")
```

The above code will produce the following output:

```
MagicFile/0274-magic-file.swift
/Users/example/Desktop/0274-magic-file.swift
Fatal error: Something bad happened!: file MagicFile/0274-magic-file.swift, line 3
```

Swift 5.8 also includes _conditional attributes_ to reduce the maintenance cost of libraries that support multiple Swift tools versions. `#if` checks can now surround attributes on a declaration, and a new `hasAttribute(AttributeName)` conditional directive can be used to check whether the compiler version has support for the attribute with the name `AttributeName` in the current language mode:

```swift
#if hasAttribute(preconcurrency)
@preconcurrency
#endif
protocol P: Sendable { ... }
```

> Swift 5.8 brings other language and standard library enhancements, including unboxing for `any` arguments to optional parameters, enabling local wrapped properties in result builders, a new `@backDeployed(before:)` attribute to back-deploy functions on older OS versions, improved debug printing for key-paths, and more.

Reviewer 5: Maybe links for each of these so that the interested user can learn more? It occurs to me that existential unboxing could merit its own blog post too: it's a complex topic and it'd be nice to have didactic material about it.

Also, in terms of what examples we list here, the `@backDeployed(before:)` attribute is really cool but (by design) something that end users should be able to ignore, so I'd argue it's not necessarily a great choice to list here—I think a user who hasn't heard about it is as likely to find it puzzling as useful ("why do I need to know about this new feature? when would I need to use it?"—you don't, basically never).

```suggestion
Swift 5.8 brings other language and standard library enhancements, including unboxing for `any` arguments to optional parameters, local wrapped properties in result builders, the `@backDeployed(before:)` attribute to back-deploy functions on older OS versions, improved debug printing for key paths, and more.
```

## Developer Experience

### Improved Result Builder Implementation

The result builder implementation has been re-worked in Swift 5.8 to greatly improve compile-time performance, code completion results, and diagnostics. The Swift 5.8 result builder implementation enforces stricter type inference that matches the semantics in [SE-0289: Result Builders](https://github.com/apple/swift-evolution/blob/main/proposals/0289-result-builders.md), which has an impact on some existing code that relied on invalid type inference.

The new implementation takes advantage of the [extended multi-statement closure inference](https://github.com/apple/swift-evolution/blob/main/proposals/0326-extending-multi-statement-closure-inference.md) introduced in Swift 5.7 and applies the result builder transformation exactly as specified by the result builder proposal - a source-level transformation which is type-checked like a multi-statement closure. Doing so enables the compiler to take advantage of all the benefits of the improved closure inference for result builder transformed code, including optimized type-checking performance (especially in invalid code) and improved error messages.

> For more details please refer to [Swift Forums Post](https://forums.swift.org/t/improved-result-builder-implementation-in-swift-5-8/63192) that outlines the improvements and provides more information about invalid inference scenarios.

Reviewer 4: `Swift Forums Post` -> the Swift Forums Post

Reviewer 5: Would lowercase "post."

Author: Thanks! I have added both suggestions in https://github.com/apple/swift-org-website/pull/257/commits/427657ddef3a52cc3f4340f9110542c7438035c0.

## Ecosystem

### Swift-DocC

> As [announced in February](https://www.swift.org/blog/tspl-on-docc/), The Swift Programming Language book has been converted to Swift-DocC and [open sourced](https://github.com/apple/swift-book), and with it came some enhancements to Swift-DocC itself in the form of [option directives](https://www.swift.org/documentation/docc/options) you can use to change the behaviour of your generated documentation. Swift-DocC has also added some new directives to create more [dynamic documentation pages](https://www.swift.org/documentation/docc/api-reference-syntax#creating-custom-page-layouts), including [Grid-based layouts](https://www.swift.org/documentation/docc/row) and [tab navigators](https://www.swift.org/documentation/docc/tab).

Reviewer 5: ```suggestion
As [announced in February](https://www.swift.org/blog/tspl-on-docc/), The Swift Programming Language book has been converted to Swift-DocC and [open sourced](https://github.com/apple/swift-book), and with it came some enhancements to Swift-DocC itself in the form of [option directives](https://www.swift.org/documentation/docc/options) you can use to change the behaviour of your generated documentation. Swift-DocC has also added some new directives to create more [dynamic documentation pages](https://www.swift.org/documentation/docc/api-reference-syntax#creating-custom-page-layouts), including [grid-based layouts](https://www.swift.org/documentation/docc/row) and [tab navigators](https://www.swift.org/documentation/docc/tab).
```

Much as I'm a fan of King's English, being Canadian myself, my understanding is that standard documentation and communication is conducted in U.S. English, yes? In which case, "behaviour," "colour," "customise," "customisation" would need to be re-spelled.

Author: While the blog is a bit more relaxed when it comes to styling (as we want to encourage authors to write in their own voice), I think a compilation of content like this—unless each section of content has a distinct author—should be consistent in styling across all contributions.

Therefore, I will change these as you've suggested. Thank you!

With the Swift 5.8-aligned release of [SwiftSyntax](https://github.com/apple/swift-syntax), SwiftSyntax contains a completely re-written parser that is implemented entirely in Swift instead of relying on the C++ parser to produce a SwiftSyntax tree. While the Swift compiler still uses the old parser implemented in C++, the eventual goal is to replace the old parser entirely. The new parser has a number of advantages:

- Contributing to or depending on SwiftSyntax is now as easy as any other Swift Package Manager package. This greatly lowers the barrier of entry for new contributors and adopters.
- The new parser has been designed with error recovery as a primary goal. It is more tolerant of parsing errors and produces better error messages.
- SwiftSyntaxBuilder allows generating source code in a declarative way using a mixture of result builders and string interpolation. An example can be found [here](https://github.com/apple/swift-syntax/blob/release/5.8/Examples/CodeGenerationUsingSwiftSyntaxBuilder.swift).

### Windows Platform

Swift 5.8 continues the incremental improvements to the Windows toolchain. Some of the important work that has gone into this release cycle includes:

> - The Windows toolchain has reduced some of the dependency on the environment variables. `DEVELOPER_DIR` was previously needed to locate components and this is no longer required. This cleans up the installation and enables us to get closer to per-user installation.

Author: 👍 Updated in https://github.com/apple/swift-org-website/pull/257/commits/b141bb076efdd4bfff96e4dca2103e16aca688f6.

- SE-0365: [Allow implicit self for weak self captures, after self is unwrapped](https://github.com/apple/swift-evolution/blob/main/proposals/0365-implicit-self-weak-capture.md)
- SE-0367: [Conditional compilation for attributes](https://github.com/apple/swift-evolution/blob/main/proposals/0367-conditional-attributes.md)
- SE-0368: [StaticBigInt](https://github.com/apple/swift-evolution/blob/main/proposals/0368-staticbigint.md)
- SE-0369: [Add CustomDebugStringConvertible conformance to AnyKeyPath](https://github.com/apple/swift-evolution/blob/main/proposals/0369-add-customdebugdescription-conformance-to-anykeypath.md)
- SE-0370: [Pointer Family Initialization Improvements and Better Buffer Slices](https://github.com/apple/swift-evolution/blob/main/proposals/0370-pointer-family-initialization-improvements.md)
- SE-0372: [Document Sorting as Stable](https://github.com/apple/swift-evolution/blob/main/proposals/0372-document-sorting-as-stable.md)
- SE-0373: [Lift all limitations on variables in result builders](https://github.com/apple/swift-evolution/blob/main/proposals/0373-vars-without-limits-in-result-builders.md)

> - SE-0374: [Add sleep(for:) to Clock](https://github.com/apple/swift-evolution/blob/main/proposals/0374-clock-sleep-for.md)

Reviewer 9: ```suggestion
```

SE-0374 is not present in Apple's 5.8 release and we should simply omit it here to avoid potential confusion.

Author: Would it be possible to learn about this by looking at the [Swift 5.8 evolutions list](https://www.swift.org/swift-evolution/#?version=5.8)?

SE-0374 is listed there for Swift 5.8, and we're referencing the list in the blog post, so the availability of this feature might be confusing.

- [SE-0378](https://github.com/apple/swift-evolution/blob/main/proposals/0378-package-registry-auth.md): Token authentication when interacting with a package registry is now supported. The `swift package-registry` command has two new subcommands `login` and `logout` for adding/removing registry credentials.

- Exposing an executable product that consists solely of a binary target that is backed by an artifact bundle is now allowed. This allow vending binary executables as their own separate package, independently of the plugins that are using them.

- In packages using tools version 5.8 or later, Foundation is no longer implicitly imported into package manifests. If Foundation APIs are used, the module needs to be imported explicitly.

See the [Swift Package Manager changelog](https://github.com/apple/swift-package-manager/blob/main/CHANGELOG.md#swift-58) for the complete list of changes.

### SwiftSyntax

> With the Swift 5.8-aligned release of [SwiftSyntax](https://github.com/apple/swift-syntax), SwiftSyntax contains a completely re-written parser that is implemented entirely in Swift instead of relying on the C++ parser to produce a SwiftSyntax tree. While the Swift compiler still uses the old parser implemented in C++, the eventual goal is to eventually replace the old parser entirely. The new parser has a number of advantages:

Reviewer 7: "the eventual goal is to eventually replace the old parser entirely"

too many 'eventually'. Maybe just "the eventual goal is to replace the old parser entirely."

Author: Great suggestion! Has been changed in 4992820.

### Windows Platform

> Swift 5.8 continues the incremental improvements to the Windows toolchain. The following tries to highlight some of the important work that has gone into this release cycle.

Reviewer 10: ```suggestion
Swift 5.8 continues the incremental improvements to the Windows toolchain. Some of the important work that has gone in this release cycles includes:
```

Author: I believe the correct grammar would be the following (based on Grammarly's suggestions):

- in -> into
- cycles -> cycle

```suggestion
Swift 5.8 continues the incremental improvements to the Windows toolchain. Some of the important work that has gone into this release cycle includes:
```
